src/components/UserForm/UserForm.jsx

```diff
@@ -0,0 +1,223 @@
+  const emails = useMemo(() => users.map(u => u.email), [users]);
+  const dispatch = useDispatch();
+  const history = useHistory();
+  const formEditStep = history.location.state?.formEditStep;
+
+  const steps = Children.toArray(children);
+
+  const getStorageValues = () => {
+    const values = JSON.parse(localStorage.getItem("values"));
+    return { ...values, birthDate: new Date(values.birthDate) };
+  };
+
+  const [continueForm, setContinueForm] = useState(() => {
+    const initial = {
+      ...initialValues,
+      step: 0,
+    };
+    const storage = getStorageValues();
+    if (!storage) return false;
+
+    const isSymmetric = isEqual(getStorageValues(), initial);
+    const showPrompt = Boolean(localStorage.getItem("values")) && isSymmetric;
+    return !showPrompt;
+  });
+
+  const [step, setStep] = useState(() => {
+    if (formEditStep || formEditStep === 0) return formEditStep;
+    if (localStorage.getItem("values") && getStorageValues()?.step)
+      return getStorageValues().step;
+
+    return 0;
+  });
+
+  const currentStep = steps[step];
+
+  const getInitialValues = () => {
+    if (valuesToEdit)
+      return { ...valuesToEdit, birthDate: new Date(valuesToEdit.birthDate) };
+
+    const inStorage = localStorage.getItem("values");
+    return inStorage ? getStorageValues() : initialValues;
+  };
+
+  const getValidationScema = useMemo(
+    () =>
+      validationSchema({
+        usernames: valuesToEdit ? [] : usernames,
+        emails: valuesToEdit ? [] : emails,
+        step,
+      }),
+    [usernames, emails, step, valuesToEdit]
+  );
+  const formProps = {
+    initialValues:
+      continueForm && !valuesToEdit ? initialValues : getInitialValues(),
+    enableReinitialize: true,
+    validationSchema: getValidationScema,
+    validateOnChange: false,
+    onSubmit: async (values, actions) => {
+      if (valuesToEdit) {
+        const payload = {
+          user: {
+            ...values,
+            birthDate: new Date(values.birthDate).getTime(),
+            lastUpdate: Date.now(),
+          },
+          key: userKey,
+          history,
+        };
+
+        dispatch(updateUserAsync(payload));
+      } else if (isLastStep() && !valuesToEdit) {
+        const payload = {
+          ...values,
+          birthDate: new Date(values.birthDate).getTime(),
+        };
+        dispatch(addUserAsync(payload));
+        stepRef.current = 0;
+        valuesRef.current = initialValues;
+        localStorage.clear();
+        history.push("/users");
+      } else {
+        stepForward();
+      }
+    },
+  };
+
+  const formik = useFormik(formProps);
+  const handleReset = () => {
+    formik.resetForm();
+    localStorage.clear();
+    setContinueForm(false);
+    setStep(0);
+  };
+  const handleContinue = () => {
+    const newStep = getStorageValues().step;
+    setContinueForm(false);
+    setStep(newStep);
+    touched.current = newStep;
+  };
+
+  const valuesRef = useRef(formik.values);
```

Comment:
why would you need values ref when you can always access it directly from formik?

Reply:
to write to the Storage only the recent values once on unmount
If I use formik.values I need to add them as a dependency which triggers my Effect function on each keystroke.
 
![Untitled](https://user-images.githubusercontent.com/46814181/109278887-1dde4e80-782a-11eb-936b-4b5d5795dae6.png)